x/mongo/driver/topology/sdam_monitoring_spec_test.go

```diff
@@ -0,0 +1,331 @@
+		WithSeedList(func(...string) []string { return cs.Hosts }),
+		WithServerOptions(func(...ServerOption) []ServerOption { return serverOptions }),
+		WithTopologySdamMonitor(func(*event.SdamMonitor) *event.SdamMonitor { return sdam }),
+	)
+
+	top, err := New(options...)
+	assert.Nil(t, err, "error creating new topology\n")
+
+	setUpServers(t, top)
+
+	return top
+}
+
+func compareServerDescriptions(t *testing.T,
+	e serverDescription, a event.ServerDescription) {
+	assert.Equal(t, e.Address, string(a.Address),
+		"expected server address %s, got %s", e.Address, a.Address)
+
+	assert.Equal(t, len(e.Hosts), len(a.Hosts),
+		"expected %d hosts, got %d", len(e.Hosts), len(a.Hosts))
+	for idx, eh := range e.Hosts {
+		ah := a.Hosts[idx]
+		assert.Equal(t, eh, string(ah), "expected host %s, got %s", eh, ah)
+	}
+
+	assert.Equal(t, len(e.Passives), len(a.Passives),
+		"expected %d hosts, got %d", len(e.Passives), len(a.Passives))
+	for idx, ep := range e.Passives {
+		ap := a.Passives[idx]
+		assert.Equal(t, ep, string(ap), "expected passive %s, got %s", ep, ap)
+	}
+
+	assert.Equal(t, e.Primary, string(a.Primary),
+		"expected primary %s, got %s", e.Primary, a.Primary)
+	assert.Equal(t, e.SetName, a.SetName,
+		"expected set name %s, got %s", e.SetName, a.SetName)
+
+	// PossiblePrimary is only relevant to single-threaded drivers.
+	if e.Type == "PossiblePrimary" {
+		e.Type = "Unknown"
+	}
+	assert.Equal(t, e.Type, a.Kind.String(),
+		"expected server kind %s, got %s", e.Type, a.Kind.String())
+}
+
+func compareTopologyDescriptions(t *testing.T,
+	e topologyDescription, a event.TopologyDescription) {
+	assert.Equal(t, e.TopologyType, a.Kind.String(),
+		"expected topology kind %s, got %s", e.TopologyType, a.Kind.String())
+	assert.Equal(t, len(e.Servers), len(a.Servers),
+		"expected %d servers, got %d", len(e.Servers), len(a.Servers))
+
+	for idx, es := range e.Servers {
+		as := a.Servers[idx]
+		compareServerDescriptions(t, es, as)
+	}
+
+	assert.Equal(t, e.SetName, a.SetName,
+		"expected set name %s, got %s", e.SetName, a.SetName)
+}
+
+func compareEvents(t *testing.T, events []monitoringEvent) {
+	var emptyUUID uuid.UUID
+
+	for idx, me := range events {
+
+		if me.TopologyOpeningEvent != nil {
+			actual, ok := publishedEvents[idx].(event.TopologyOpeningEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.TopologyOpeningEvent{}, actual)
+			assert.False(t, uuid.Equal(uuid.UUID(actual.ID), emptyUUID), "expected topology id")
+		}
+		if me.ServerOpeningEvent != nil {
+			actual, ok := publishedEvents[idx].(event.ServerOpeningEvent)
+			assert.True(t, ok, "expected type %T, got %T", event.ServerOpeningEvent{}, actual)
+
+			evt := me.ServerOpeningEvent
+			assert.Equal(t, evt.Address, string(actual.Address),
+				"expected address %s, got %s", evt.Address, actual.Address)
+			// TODO: assert.False(t, uuid.Equal(uuid.UUID(actual.ID), emptyUUID), "expected topology id %v", actual.ID)
```

Comment:
Currently in the spec tests the expected topology ID will just be a dummy value of 42. It seems that it would be better to not compare the actual topology ID against this value, and to just check if it is a non-nil value instead, hence why this is commented out at the moment. It can be changed to check for a non-nil topology ID, or omitted from the test entirely?